src/hotspot/share/compiler/compileBroker.cpp

```diff
@@ -367,33 +367,31 @@ void CompileQueue::add(CompileTask* task) {
  */
 void CompileQueue::delete_all() {
   MutexLocker mu(MethodCompileQueue_lock);
-  CompileTask* next = _first;
+  CompileTask* current = _first;
 
   // Iterate over all tasks in the compile queue
-  while (next != nullptr) {
-    CompileTask* current = next;
-    next = current->next();
-    bool found_waiter = false;
-    {
-      MutexLocker ct_lock(CompileTaskWait_lock);
-      assert(current->waiting_for_completion_count() <= 1, "more than one thread are waiting for task");
-      if (current->waiting_for_completion_count() > 0) {
-        // If another thread waits for this task, we must wake them up
-        // so they will stop waiting and free the task.
-        CompileTaskWait_lock->notify_all();
-        found_waiter = true;
-      }
-    }
-    if (!found_waiter) {
-      // If no one was waiting for this task, we need to delete it ourselves.
-      // In this case, the task is also certainly unlocked, because, again, there is no waiter.
-      // Otherwise, by convention, it's the waiters responsibility to delete the task.
+  while (current != nullptr) {
+    if (!current->is_blocking()) {
+      // Non-blocking task. No one is waiting for it, delete it now.
       delete current;
+    } else {
+      // Blocking task. By convention, it is the waiters responsibility
+      // to delete the task. We cannot delete it here, because we do not
+      // coordinate with waiters. We will notify the waiters later.
     }
+    current = current->next();
   }
   _first = nullptr;
   _last = nullptr;
 
+  // Wake up all blocking task waiters to delete all remaining blocking
+  // tasks. This is not a performance sensitive path, so we do this
+  // unconditionally to simplify coding.
+  {
+    MonitorLocker ml(Thread::current(), CompileTaskWait_lock);
+    ml.notify_all();
+  }
+
```

Comment:
What about other compiler threads which still in process of compiling for blocking tasks? They still need it CompileTask object.
`delete_all()` is called by one compiler thread which finished compilation but other threads may not.

I don't see any compiler thread checks `shut_down` state to stop compilation.

Reply:
Ah, your question is what happens if we notify here, and compilations are still running? Well, I think current protocol should nominally allow waiters to wait until compilation is over and then allow them to delete the task. But then I see `wait_for_compilation` can exit when compilation is shut down:

```
    while (!task->is_complete() && !is_compilation_disabled_forever()) {
      ml.wait();
    }
```

This will delete the task while compiler thread is running. Grrr.

Reply:
Can compiler thread delete its **own** blocking task when he finished. And let Java thread resume execution when compilation disabled as it do now but do nothing about task in such case?